design/endpoint-slice-support.md

```diff
@@ -0,0 +1,101 @@
+# Contour k8s endpoint slice support
+
+Contour currently supports Kubernetes endpoints.
+This document proposes a process to enable support for k8s endpoint slices and, eventually, migrate to exclusively using it.
+
+## Goals
+
+- Contour will support k8s endpoint slices.
+- The endpoint slices will be gated by a feature flag in the Contour config.
+- Kubernetes endpoints will still be the default mode of operation.
+
+## Non-goals
+
+- Removal of k8s endpoints.
+- Support advanced features such as topology aware routing/hints offered by k8s endpoints.
+
+## Background
+
+See [#2696](https://github.com/projectcontour/contour/issues/2696)
+
+Before the introduction of the endpoint slices, the only method for a service to expose the IP addresses and ports of the pads it fronted was through Kubernetes endpoints.
+
+However, Kubernetes endpoints have a few challenges with scaling, such as
+- transmission of the entire endpoint object even if one endpoint changes, leading to network contention and heavy recalculation on the consumers of this data.
+- As of k8s v1.22, a hard limit of 1000 addresses per endpoint object.
+- No topology aware information for more intelligent routing.
+
+To solve the above problems, endpoint slices was introduced in 1.16 and, as of k8s v1.21 in [stable state](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+
+## High-Level Design
+
+A feature flag will gate the usage of endpoint slices, and the implementation will use the same pattern as used in the `endpointtranslator.go` with a few differences listed in the detailed design.
+
+## Detailed Design
```

Comment:
Could you add some info about how [FQDN address types](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#address-types) are handled? Off the top of my head, I think we'd want to handle them similarly to ExternalName services (i.e. disabled by default since there are some security concerns), but would need to look at it a little more to be sure.

Reply:
So, I looked into this and am leaning towards not supporting it. 

To use the endpointSlice FQDN address type, a user needs to create the EndpointSlice object manually and attach it to a service via labels, something like this 

```
apiVersion: v1
kind: Service
metadata:
  name: fqdn-test
spec:
  ports:
    - name: http
      port: 80
      targetPort: 80
---
apiVersion: discovery.k8s.io/v1
kind: EndpointSlice
metadata:
  name:  fqdn-test
  labels:
    kubernetes.io/service-name: fqdn-test
ports:
  - name: http
    appProtocol: http
    protocol: TCP
    port: 80
endpoints:
  - addresses:
      - "projectocontour.io"
```

While the [PR](https://github.com/projectcontour/contour/pull/5745) I have open should work with this out of the box, 

I have the below reasoning to disable it explicitly. 
- I think there is possibly a small set use-cases that this solves that externalName services don't solve. 
- Limit the scope of features to avoid the introduction of confounding variables. 

wdyt?